File: 2017-JS/Day-2/day2.js
const fs = require('fs');

let input = fs.readFileSync('input.txt','utf-8').split('\r');
let lines = [];

for(let i = 0; i < input.length; i++){
    let line = input[i].split('\t');
    //line[line.length-1] = line[line.length - 1].slice(0, line[line.length-1].length - 1); if split by \n
    lines.push(line.map((element) => Number(element)));
}

Part1(lines);
Part2(lines);

function Part1(lines){
    let toR = 0;
    for(let i = 0; i < lines.length; i++){
        let numLine = lines[i];   
        let res = getDiffOfMaxAndMin(numLine);
        toR += res;
    }
    console.log(toR);
}
function Part2(lines){
    let result = 0;
    for(let i = 0; i < lines.length; i++){
        let divisorFound = false;
        let numLine = lines[i];
        for(let j = 0; j < numLine.length; j++){
            for(let k = 0; k < numLine.length; k++){
                if(k != j && numLine[j] % numLine[k] == 0){
                    divisorFound = true;
                    result += numLine[j] / numLine[k];
                }
                if(divisorFound){break;}
            }
            if(divisorFound){break;}
        }
        
    }
    console.log(result);
}

function getDiffOfMaxAndMin(input){
    //console.log(input);
    let min = 999999999;
    let max = -999999999;
    for(let i = 0; i < input.length; i++){
        let nr = input[i];
        if(nr < min){ min = nr;}
        if(nr > max){ max = nr;}
    }

    return max - min;
}

File: 2016-Dart/Day-2/day2.dart
import 'dart:convert';
import 'dart:ffi';
import 'dart:io';

void main(){
  var file = File('./input.txt');
  List<String> lines = file.readAsStringSync(encoding: utf8).split('\r\n');
  part1(lines);
  part2(lines);
}

void part1(List<String> lines){
  List<String> toR = [];
  List<List<int>> grid = [[1,2,3], [4,5,6], [7,8,9]];
  int ci = 1, cj = 1;
  for(final line in lines){
    for(int i = 0; i < line.length; i++){
      if(line[i] == 'U'){
        if(ci-1 >= 0){ci = ci-1;}
      }else if(line[i] == 'R'){
        if(cj+1 < grid[0].length){cj = cj+1;}
      }else if(line[i] == 'D'){
        if(ci+1 < grid.length){ci = ci+1;}
      }else if(line[i] == 'L'){
        if(cj-1 >= 0){cj = cj-1;}
      }
    }
  toR.add(grid[ci][cj].toString());
  }
  print(toR.join());
}

void part2(List<String> lines){
  List<String> toR = [];
  List<List<String>> grid = [['-1', '-1', '1', '-1', '-1'], ['-1','2','3','4','-1'], ['5', '6','7', '8', '9'], ['-1', 'A', 'B', 'C', '-1'], ['-1','-1','D','-1','-1']];
  int ci = 2, cj = 0;
  for(final line in lines){
    for(int i = 0; i < line.length; i++){
      if(line[i] == 'U'){
        if(ci-1 >= 0 && grid[ci-1][cj] != '-1'){ci = ci-1;}
      }else if(line[i] == 'R'){
        if(cj+1 < grid[0].length && grid[ci][cj+1] != '-1'){cj = cj+1;}
      }else if(line[i] == 'D'){
        if(ci+1 < grid.length && grid[ci+1][cj] != '-1'){ci = ci+1;}
      }else if(line[i] == 'L'){
        if(cj-1 >= 0 && grid[ci][cj-1] != '-1'){cj = cj-1;}
      }
    }
  toR.add(grid[ci][cj]);
  }
  print(toR.join());
}